lib/ansible/modules/cloud/misc/qemu_img.py

```diff
@@ -0,0 +1,221 @@
+        argument_spec = dict(
+            dest = dict(type='str', required=True),
+            options = dict(type='str', default='preallocation=metadata'),
+            format = dict(type='str', default='qcow2'),
+            size = dict(type='str'),
+            grow= dict(type="bool", default=True),
+            shrink = dict(type="bool", default=False),
+            state = dict(type='str', choices=['absent', 'present'], default='present'),
+        ),
+        supports_check_mode = True,
+    )
+
+    result = dict(
+        changed = False,
+    )
+
+    state = module.params['state']
+    dest = module.params['dest']
+    img_format = module.params['format']
+    img_options = module.params['options']
+    size = module.params['size']
+    grow = module.boolean(module.params['grow'])
+    shrink = module.boolean(module.params['shrink'])
+
+    qemu_img = module.get_bin_path('qemu-img', True)
+
+    if state == 'present':
+        if not size:
+            module.fail_json(msg="No size defined, creating a disk image requires a size")
+        if not os.path.exists(dest):
+            if not module.check_mode:
+                if not img_options:
+                    module.run_command('%s create -f %s "%s" %s' % (qemu_img, img_format, dest, size), check_rc=True)
+                else:
+                    module.run_command('%s create -f %s -o "%s" "%s" %s' % (qemu_img, img_format, img_options, dest, size), check_rc=True)
+            result['changed'] = True
+        else:
+            size_unit = size[-1:]
+            if size_unit not in ['T','G','M','k','b','']:
+                module.fail_json(msg="No valid size unit specified")
+            else:
+                if(size_unit == 'T'):
+                    unitmultiplier = 1024*1024*1024*1024
+                elif(size_unit == 'G'):
+                    unitmultiplier = 1024*1024*1024
+                elif(size_unit == 'M'):
+                    unitmultiplier = 1024*1024
+                elif(size_unit == 'k'):
+                    unitmultiplier = 1024
+                else:
+                    unitmultiplier = 1
+            size = size[:-1]
+            size_increase = False
+            size_decrease = False
+            if(size[:1] == '+'):
+                size_increase = True
+                size = size[1:]
+            elif(size[:1] == '-'):
+                size_descrease = True
+                size = size[1:]
+            size = int(float(size)*float(unitmultiplier))
+            rc, stdout, _ = module.run_command('%s info "%s"' % (qemu_img, dest), check_rc=True)
+            current_size = None
+            for line in stdout.splitlines():
+                if 'virtual size' in line:
+                    ### virtual size: 5.0M (5242880 bytes)
+                    current_size = int(line.split('(')[1].split()[0])
+            if not current_size:
+                module.fail_json(msg='Unable to read virtual disk size of %s' % (dest))
+            if(grow):
```

Comment:
The grow and shrink booleans are used for allowing to grow and shrink the image. (That is why they default resp. to true and false). So you have to compare the original size against the wanted size, and if it shrinks and shrinking is allowed, only then should it resize (otherwise it needs to escalate). Similar for the grow-case.